independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java

```diff
@@ -101,32 +86,34 @@ public void destroy(Contextual<?> contextual) {
         }
     }
 
-    private void fireIfNotEmpty(LazyValue<EventImpl.Notifier<Object>> value) {
-        EventImpl.Notifier<Object> notifier = value.get();
-        if (!notifier.isEmpty()) {
-            notifier.notify(toString());
+    @Override
+    public void activate(ContextState initialState) {
+        if (initialState == null) {
+            currentContext.set(new ConcurrentHashMap<>());
+        } else {
+            if (initialState instanceof RequestContextState) {
+                currentContext.set(((RequestContextState) initialState).value);
+            } else {
+                throw new IllegalArgumentException("Invalid inital state: " + initialState);
+            }
         }
+        // Fire an event with qualifier @Initialized(RequestScoped.class) if there are any observers for it
+        fireIfNotEmpty(initializedNotifier);
```

Comment:
I did wonder about this: is it good that we're firing notifiers for every context restore? I've never used `@Initialized` but are they expected to be called more than once for a single request context?

Reply:
That's a good point. I think we should only fire the event in case of `initialState == null`. The spec says that we should fire the event _"when the request context is initialized"_. But `initialized` is not well-defined. I'll change this...